test.py

```diff
@@ -1,11 +1,74 @@
 from langchain.chat_models import AzureChatOpenAI
 from dotenv import load_dotenv
+import os
+import streamlit as st
+import requests
+import en_core_web_sm
+import geocoder
 
-# Load environment variables
-load_dotenv()
 
-# Test GPT-3
-gpt_turbo = AzureChatOpenAI(deployment_name="gpt-turbo", temperature=0)
-prompt = "What is the weather today in Madrid?"
-gpt3_answer = gpt_turbo.predict(prompt)
-print(gpt3_answer)
\ No newline at end of file
+# I was not able to get the key to work loading from the env why?
+API_WEATHER_KEY = os.getenv("API_WEATHER_KEY")
+
+
+# Find name of a city in a prompt
+def find_city(input_text):
+    # Load the spaCy model
+    nlp = en_core_web_sm.load()
+    doc = nlp(input_text)
+
+    # if the city name is in the data_base & has the location prefix return the city name:
+    for entity in doc.ents:
+        if entity.label_ == 'GPE':
+            return entity.text
+    else:
+        return None
+
+
+# Find coordinates of a city
+def find_coordinates(city_name):
+    geo_data = requests.get(
+        f"https://api.openweathermap.org/geo/1.0/direct?q={city_name}&limit=10&appid={API_WEATHER_KEY}&units").json()
+    return geo_data[0]["lat"], geo_data[0]["lon"]
+
+
+# Main function
+def run_luna():
+    # Load environment variables (if needed)
+    load_dotenv()
+
+    st.title('🌚 Welcome to Luna')
+
+    # creating a text box
+    with st.form('my_form'):
+        # waiting for user input
+        prompt = st.text_area('Enter a question: ')
+        submitted = st.form_submit_button('Submit')
+
+        # if prompt is submitted
+        if submitted:
+
+            # responding to weather questions
+            if "weather" in prompt:
```

Comment:
In order to extract the details of the weather that are relevant for the user query, we could do a preprocessing through the LLM. For example:

```python
llm_prompt = f"""
Instruction: Extract details relevant to respond a weather question by the user...
# In-context-learning examples to show variables we want it to extract (location, time, ...) and the format of the answer (so that the output variables can be parsed later
# Task n
User_query: ---
Answer:
"""
```